packages/core-mobile/app/seedless/services/SeedlessSession.ts

```diff
@@ -121,101 +116,68 @@ class SeedlessSessionManager {
       mfaReceipt
     )
 
-    const sessionResponse = signResponse.requiresMfa()
-      ? signResponse.mfaSessionInfo()
-      : signResponse.data()
-
-    if (sessionResponse) {
-      await SignerSessionManager.createFromSessionInfo(
-        envInterface,
-        SEEDLESS_ORG_ID,
-        sessionResponse,
-        this.sessionStorage
-      )
+    if (response.requiresMfa()) {
+      // if MFA is required, we cannot access session data until MFA verification is complete
+      // thus, we are just storing the MFA client here (for verification operations)
+      const client = await response.mfaClient()
+      if (!client) throw new Error('MFA client is missing')
+      this.signerClient = client
+    } else {
+      // when MFA is not required or MFA verification is complete
+      // we can safely access session data
+      const sessionData = response.data()
+
+      // destroy outdated client instance
+      this.signerClient = null
+
+      // persist session data
+      await this.sessionManager.store(sessionData)
     }
 
-    return signResponse
+    return response
   }
 
-  async refreshToken(): Promise<Result<void, TokenRefreshErrors>> {
+  async refreshToken(): Promise<Result<void, Error>> {
     this.setIsTokenValid(false)
 
-    const sessionMgr = await SignerSessionManager.loadFromStorage(
-      this.sessionStorage
-    ).catch(reason => {
-      Logger.error('Failed to load session manager from storage', reason)
-      return undefined
-    })
+    try {
+      await this.sessionManager.forceRefresh()
+      this.setIsTokenValid(true)
+      return { success: true, value: undefined }
+    } catch (err) {
+      // the forceRefresh method does not emit a token expired event
+      // so we need to catch it manually here
+      if (
+        err instanceof ErrResponse &&
+        err.status === 403 &&
+        err.isSessionExpiredError()
+      ) {
+        this.onSessionExpired?.()
+      }
 
-    if (!sessionMgr) {
       return {
         success: false,
-        error: {
-          name: 'RefreshFailed',
-          message: 'Failed to load session manager from storage'
-        }
+        error: err as Error
       }
     }
-    const refreshResult = await retry({
```

Comment:
also we don't need to do retry since internally cubist already does retry for 500 http errors.